### PR TITLE
add support for upload lifetime

### PR DESCRIPTION
**Pull request by dansan**
- add support for upload lifetime
- add '-h' for '--help'
- fix falling back type output

Clicks _Tuples as Multi Value Options_ feature needs higher clicks-version. 4.0 would have been enough it seems: [http://click.pocoo.org/5/options/#tuples-as-multi-value-options](http://click.pocoo.org/5/options/#tuples-as-multi-value-options) but my pip simply updated to 6.6. Don't know if that's OK...

### CODE REVIEW

File: bepasty_cli/cli.py [hunk 1]
```diff
@@ -125,8 +151,8 @@ def upload(token, filename, fname, url, ftype, insecure):
         ftype = mime.from_buffer(first_chunk).decode()
 
         if not ftype:
-            print('falling back to {}'.format(ftype))
             ftype = 'text/plain'
+            print('falling back to {}'.format(ftype))
```
Comment (ThomasWaldmann) [inline bepasty_cli/cli.py]: maybe don't do unrelated changes. but rather keep them for a separate PR.

File: bepasty_cli/cli.py [hunk 2]
```diff
@@ -55,12 +76,17 @@
     '--insecure',
     help='Disable SSL certificate validation',
     is_flag=True)
-def main(token, filename, fname, url, ftype, list_pastes, insecure):
+def main(token, filename, fname, url, ftype, list_pastes, insecure, max_lifetime):
     url = url.rstrip("/")
+    max_lifetime = (max_lifetime[0], LIFETIME_MAPPING[max_lifetime[1]])
+    if max_lifetime[0] < 1:
+        print('Invalid "max-lifetime" value "{}". Setting max-lifetime to "forever".'.format(max_lifetime[0]))
+        max_lifetime = (1, 'FOREVER')
```
Comment (ThomasWaldmann) [inline bepasty_cli/cli.py]: that's a bit unusual - usually a tool would rather fail if it gets invalid parameters.

File: bepasty_cli/cli.py [hunk 3]
```diff
@@ -156,6 +182,8 @@ def upload(token, filename, fname, url, ftype, insecure):
             'Content-Type': ftype,
             'Content-Filename': fname,
             'Content-Length': str(len(payload)),  # rfc 2616 14.16
+            'Maxlife-Unit': max_lifetime[1],
+            'Maxlife-Value': str(max_lifetime[0])
```
Comment (ThomasWaldmann) [inline bepasty_cli/cli.py]: add a comma at the end, see line 184.

**Comment by ThomasWaldmann** [comment]
about click version:
- setup.py installs the latest click anyway
- requirements.txt should just have some tested versions (for click, for the other stuff).

File: bepasty_cli/cli.py [hunk 4]
```diff
@@ -38,6 +50,15 @@
     'fname',
     help='Filename for piped input.')
 @click.option(
+    '-m',
```
Comment (dansan) [inline bepasty_cli/cli.py]: Sorry - I don't understand. Both are in use:
```
-l, --list                      Lists all pastes on server (requires LIST
-t, --type TEXT                 Filetype for piped input. Give the value for
```

File: bepasty_cli/cli.py [hunk 5]
```diff
@@ -38,6 +50,15 @@
...
+    '--max-lifetime',
+    type=(int, click.Choice(LIFETIME_CHOICES)),
+    default=(1, 'f'),
+    help='Lifetime for the file that is uploaded. Example: "-m 2 d" (two days). If this '
+         'option is not set, the uploads lifetime is "forever". Multiplier and unit are '
+         'space separated. Unit has to be one of these: "min" (minutes), "h" (hours), '
```
Comment (dansan) [inline bepasty_cli/cli.py]: Yes, that's right. But this way we can use an existing declarative programming deature of clicks.
If we'd go with `-m 10d` we'd have to parse and error-message ourselfs. I think the usual (w/o space) style stems from the complexity of traditional cmdline parsing.

If you think it's important, I would try to code a [custom clicks type](http://click.pocoo.org/5/parameters/#implementing-custom-types).

File: bepasty_cli/cli.py (hunk 2 above)
Comment (dansan) [inline bepasty_cli/cli.py]: Ah yes. It is meant to mimic the behavior of the server (which switches without raising an error to forever for all wrong lifetime values).
But I concur, that this is not what you'd expect from a cmdline tool and will change it.

File: bepasty_cli/cli.py (hunk 1 above)
Comment (dansan) [inline bepasty_cli/cli.py]: Sorry, lazy ;)
Will do.

File: bepasty_cli/cli.py (hunk 3 above)
Comment (dansan) [inline bepasty_cli/cli.py]: ACK

File: bepasty_cli/cli.py (hunk 4 above)
Comment (ThomasWaldmann) [inline bepasty_cli/cli.py]: well, using m (and not f) just because the setting is max-foo is not very intuitive.
ok, as the more intuitive (lowercase) options in this case are already taken, -L or -T maybe?

File: bepasty_cli/cli.py (hunk 2 above)
Comment (ThomasWaldmann) [inline bepasty_cli/cli.py]: ok, thanks.

File: bepasty_cli/cli.py (hunk 5 above)
Comment (ThomasWaldmann) [inline bepasty_cli/cli.py]: @NotANormalNerd your opinion?

**Comment by dansan** [comment]
It's surprisingly simple to write  a custom clicks argument :)

It accepts both `-m 1d` and` -m '1 d'` (and also `day` instead of `d`).

File: bepasty_cli/cli.py [hunk 6]
```diff
@@ -30,6 +31,18 @@
 LIFETIME_MAPPING = dict(zip(LIFETIME_CHOICES, LIFETIME_NAMES))
 
 
+class LifetimeParamType(click.ParamType):
+    name = 'lifetime'
+    lifetime_regex = re.compile(r'|'.join(r'^([1-9]\d*?) ?({})$'.format(choice) for choice in LIFETIME_CHOICES))
```
Comment (ThomasWaldmann) [inline bepasty_cli/cli.py]: while giving "099d" would be a bit strange, we do not need to disallow leading zeros, do we?

so, just use `(\d+)` there?

Comment (ThomasWaldmann) [inline bepasty_cli/cli.py]: hmm, we just need to disallow that someone gives e.g. `0d`, as giving a zero time (no matter what unit) is invalid.

but this maybe does not need to get addressed by the regex, can be checked later.

File: bepasty_cli/cli.py [hunk 7]
```diff
@@ -30,6 +31,18 @@
...
+
+    def convert(self, value, param, ctx):
+        m = self.lifetime_regex.match(value)
+        if m:
+            return [x for x in m.groups() if x]
```
Comment (ThomasWaldmann) [inline bepasty_cli/cli.py]: why `if x`?

File: bepasty_cli/cli.py [hunk 8]
```diff
@@ -50,13 +63,13 @@
...
 @click.option(
-    '-m',
-    '--max-lifetime',
-    type=(int, click.Choice(LIFETIME_CHOICES)),
-    default=(1, 'f'),
-    help='Lifetime for the file that is uploaded. Example: "-m 2 d" (two days). If this '
-         'option is not set, the uploads lifetime is "forever". Multiplier and unit are '
-         'space separated. Unit has to be one of these: "min" (minutes), "h" (hours), '
+    '-L',
+    '--lifetime',
```
Comment (ThomasWaldmann) [inline bepasty_cli/cli.py]: better :)

File: bepasty_cli/cli.py (hunk 7 above)
Comment (dansan) [inline bepasty_cli/cli.py]: m.group() returns `None` for each unmatched regex-group (int, 'min', int, 'd', int, 'day', int, 'm', int, 'month',...)
`-m 2d` → [None, None, 2, 'd', None, None, None, ...]

File: bepasty_cli/cli.py [hunk 9]
```diff
@@ -33,14 +33,18 @@
 
 class LifetimeParamType(click.ParamType):
     name = 'lifetime'
-    lifetime_regex = re.compile(r'|'.join(r'^([1-9]\d*?) ?({})$'.format(choice) for choice in LIFETIME_CHOICES))
+    lifetime_regex = re.compile(r'|'.join(r'^(\d+?) ?({})$'.format(choice) for choice in LIFETIME_CHOICES))
```
Comment (ThomasWaldmann) [inline bepasty_cli/cli.py]: why is it `\d+?` and not `\d+` in that term?

File: bepasty_cli/cli.py (hunk 7 above)
Comment (ThomasWaldmann) [inline bepasty_cli/cli.py]: ah, ok.

File: bepasty_cli/cli.py [hunk 10]
```diff
@@ -18,7 +19,35 @@
...
+    name = 'lifetime'
+    lifetime_regex = re.compile(r'|'.join(r'^(\d+?) ?({})$'.format(choice) for choice in LIFETIME_CHOICES))
+
+    def convert(self, value, param, ctx):
+        m = self.lifetime_regex.match(value)
+        if m:
+            result = [x for x in m.groups() if x is not None]
+            result[0] = result[0].lstrip('0')
+            if not result[0]:
+                self.fail('Multiplier for the lifetime argument must be positive.', param, ctx)
```
Comment (ThomasWaldmann) [inline bepasty_cli/cli.py]: hmm, sorry, I was somehow thinking about converting to int and doing int-based processing, but didn't consider we work with strings only here.

so, guess your initial version was better than this alternative.

**Comment by dansan** [comment]
> more like matching (int) (d|day|m|month|...).
No idea why I hadn't thought of that ^.^
Forever doesn't need a multiplier now - seemed silly.

File: bepasty_cli/cli.py [hunk 11]
```diff
@@ -33,15 +33,15 @@
...
     def convert(self, value, param, ctx):
         m = self.lifetime_regex.match(value)
         if m:
-            result = [x for x in m.groups() if x is not None]
-            result[0] = result[0].lstrip('0')
-            if not result[0]:
-                self.fail('Multiplier for the lifetime argument must be positive.', param, ctx)
+            result = m.groups()
+            result = result[0].lstrip('0'), result[1]
+            if not result[0] and not result[1] in ['f', 'forever']:
```
Comment (ThomasWaldmann) [inline bepasty_cli/cli.py]: x not in y

^ slightly more elegant, "not in" is a operator.

**Comment by dansan** [comment]
'-m f' produced an empty 'Maxlife-Value' in the headers, which the server cannot handle

File: bepasty_cli/cli.py [hunk 12]
```diff
@@ -38,6 +70,15 @@
     'fname',
     help='Filename for piped input.')
 @click.option(
+    '-L',
+    '--lifetime',
+    type=LifetimeParamType(),
+    default='1f',
+    help='Lifetime for the file that is uploaded. Example: "-m 2d" (two days). If this '
```
Comment (ThomasWaldmann) [inline bepasty_cli/cli.py]: Example is out of sync with short option (-L).

**Comment by ThomasWaldmann** [comment]
you mean `-L f` ? found same minor mistake in help text.

besides that, it looks good. :)

do the small fix, collapse all (or most) changesets.

**Comment by dansan** [comment]
Damn - "fix falling back type output" also ended up in the squashed commit. Should have used a separate branch from the beginning… 
Shouldn't be a problem for the merge though.

**Comment by ThomasWaldmann** [comment]
squashed? gh still shows 7 commits.

**Comment by dansan** [comment]
Oops - I squashed them into 1 together, but didn't delete the previous commits.

**Comment by ThomasWaldmann** [comment]
ok, thanks for working on that!
